src/main/scala/sangria/relay/Connection.scala

```diff
@@ -128,65 +128,54 @@ object Connection {
-    val (lastEdgeOffset, actualEndOffset) = if (0 <= beforeOffset && beforeOffset < arrayLength) {
-      val lEO = beforeOffset - 1
-      val actualEndOffset = math.min(endOffset, beforeOffset)
-      lEO -> actualEndOffset
-    } else (arrayLength - 1) -> endOffset
-
-    val numberEdgesAfterCursor = lastEdgeOffset - firstEdgeOffset + 1
-
-    val finalEndOffset =
-      first.fold(actualEndOffset)(f => math.min(actualEndOffset, f + actualStartOffset))
-    val finalStartOffset =
-      last.fold(actualStartOffset)(l => math.max(actualStartOffset, actualEndOffset - l))
-
-    // If supplied slice is too large, trim it down before mapping over it.
-    val slice = arraySlice.slice(
-      math.max(finalStartOffset - sliceStart, 0),
-      arraySlice.size - (sliceEnd - finalEndOffset))
-
-    val edges = slice.zipWithIndex.map { case (value, index) =>
-      Edge(value, offsetToCursor(finalStartOffset + index))
+    val afterOffset = getOffset(after)
+    val beforeOffset = getOffset(before)
+
+    // after based on before + last
+    val backwardPaginationAfterMaybe = (beforeOffset, last) match {
+      // (before - last) gives the index of first item to return but after cursor needs to be before it
+      case (Some(before), Some(last)) => Some(before - last - 1)
+      case (_, Some(last)) => Some(arrayLength - last - 1)
+      case _ => None
     }
-
-    val firstEdge = edges.headOption
-    val lastEdge = edges.lastOption
-
-    val hasPreviousPage = (last, after) match {
-      case (Some(l), _) => numberEdgesAfterCursor > l
-      case (None, Some(_)) => afterOffset >= 0
-      case (_, _) => false
+    val finalAfterMaybe = List(backwardPaginationAfterMaybe, afterOffset).flatten.maxOption
+
+    // before based on after + first
+    val forwardPaginationBeforeMaybe = (afterOffset, first) match {
+      // (after + first) gives the index of last item to return but before cursor needs to be after it
+      case (Some(after), Some(first)) => Some(after + 1 + first)
+      case (_, Some(first)) => Some(first)
+      case _ => None
     }
+    val finalBeforeMaybe = List(forwardPaginationBeforeMaybe, beforeOffset).flatten.minOption
 
-    val hasNextPage = (first, before) match {
-      case (Some(f), _) => numberEdgesAfterCursor > f
-      case (_, Some(_)) => beforeOffset < arrayLength
-      case (_, _) => false
+    // align slice indices with all edges indices
+    val sliceWithIdx = arraySlice.zipWithIndex.map { case (e, idx) => (e, idx + sliceStart) }
+    val trimmedSlice = sliceWithIdx.filter { case (_, idx) =>
+      finalAfterMaybe.forall(_ < idx) && finalBeforeMaybe.forall(_ > idx)
     }
 
-    DefaultConnection(
-      PageInfo(
-        hasNextPage,
-        hasPreviousPage,
-        startCursor = firstEdge.map(_.cursor),
-        endCursor = lastEdge.map(_.cursor)
-      ),
-      edges
-    )
+    // If result is empty and firstM and lastM are empty then no paging can be done
```

Comment:
```suggestion
    // If result is empty and first and last are empty then no paging can be done
```
(and before and after?)